plugins/derive/derive.ml

```diff
@@ -10,44 +10,69 @@
 
 open Context.Named.Declaration
 open Vernacentries.DefAttributes
+open Constrexpr
+
+let check_allowed_implicit ?loc = function
+  | Generalized _ | Default (MaxImplicit | NonMaxImplicit) ->
+    CErrors.user_err ?loc (Pp.str "Unexpected implicit argument annotation.")
+  | Default Explicit -> ()
+
+let check_allowed_binders = function
+  | CLocalAssum (n::_, _, impl, _) -> check_allowed_implicit ?loc:n.CAst.loc impl
+  | CLocalPattern _ -> () (* a priori acceptable, but interp_named_context_evars_as_arguments does not support it *)
+  | CLocalDef (n, _, _, _) -> ()
+  | CLocalAssum ([], _, _, _) -> assert false
+
+let rec fill_assumptions env sigma = function
+  | [] -> sigma, env, []
+  | LocalAssum (na,t) :: ctx ->
+    let sigma, ev = Evarutil.new_evar env sigma ~src:(Loc.tag @@ Evar_kinds.GoalEvar) ~typeclass_candidate:false t in
+    let decl = LocalDef (na,ev,t) in
+    let sigma, env, ctx = fill_assumptions (EConstr.push_named decl env) sigma ctx in
+    sigma, env, decl :: ctx
+  | LocalDef _ as decl :: ctx ->
+    let sigma, env, ctx = fill_assumptions (EConstr.push_named decl env) sigma ctx in
+    sigma, env, decl :: ctx
```

Comment:
I guess this could be written using List.fold_right_map
Do it if you like the idea, otherwise it's fine

Reply:
I prefer explicit recursions to combinators when the body is non-trivial, personally.